### PR TITLE
feat(vscode): If uno.root is not set,  auto-detection

### PR DESCRIPTION
**Pull request by Simon-He95**
With the popularity of monorepo, more and more libraries may use this mode, but it is very troublesome to configure `settings.json` additionally. I hope to automate changes more intelligently based on the `unocss.config` of my current project. #3106 #3034 #3009 #2742 #3000

### CODE REVIEW

File: packages/vscode/src/index.ts [hunk 1]
```diff
@@ -70,29 +69,41 @@ export async function activate(ext: ExtensionContext) {
...
+
+  const cacheMap = new Set()
+
+  const registerUnocss = async () => {
+    const url = window.activeTextEditor?.document.uri.fsPath
+    if (!url)
+      return
+    if (/node_modules/.test(url))
+      return
+    const target = await findUp(['package.json'], { cwd: url })
```
Comment (antfu) [inline packages/vscode/src/index.ts:82]: Maybe we should search for the uno.config directly instead.

File: packages/vscode/src/index.ts [hunk 2]
```diff
@@ -70,29 +69,41 @@ export async function activate(ext: ExtensionContext) {
...
+    if (!target)
+      return
+    const cwd = target.slice(0, -13)
```
Comment (antfu) [inline packages/vscode/src/index.ts:85]: Not in favor of this magic number, can we use `basename`?

File: packages/vscode/src/index.ts [hunk 3]
```diff
@@ -70,29 +69,41 @@ export async function activate(ext: ExtensionContext) {
...
-          registerAnnotations(cwd, contextLoader, status, ext)
-          registerSelectionStyle(cwd, contextLoader)
-        }
-      }
-      contextLoader.reload()
-      log.appendLine('✅ Reloaded.')
-    }),
-  )
+
+  const cacheMap = new Set()
```
Comment (antfu) [inline packages/vscode/src/index.ts:74]: We should also cache for file resolution, so we don't need to run `findUp` every time user navigate to the file.

Comment (antfu) [reply packages/vscode/src/index.ts:74]: In addition, we could also listen to the fs on adding/removing the config file and purge the cache.

File: packages/vscode/src/index.ts [hunk 4]
```diff
@@ -70,29 +69,41 @@ export async function activate(ext: ExtensionContext) {
...
+
+  const registerUnocss = async () => {
+    const url = window.activeTextEditor?.document.uri.fsPath
+    if (!url)
+      return
+    if (/node_modules/.test(url))
```
Comment (antfu) [inline packages/vscode/src/index.ts:80]: I think this might be better to be configurable, as an `exclude` option maybe.

**Review by antfu — CHANGES_REQUESTED** [review]

**Comment by antfu** [comment]
Also I don't think we should completely remove the `unocss.root` support, we could have this as a smart default when root is not specified, but keep the original behaviour without breaking change.

**Comment by Simon-He95** [comment]
> Also I don't think we should completely remove the `unocss.root` support, we could have this as a smart default when root is not specified, but keep the original behaviour without breaking change.

Thank you for your review, I made some adjustments, please help to check it out

File: test/assets/variant-group.vue [hunk 5]
```diff
@@ -1,8 +1,8 @@
 <script setup lang="ts">
 const a = 1
 const b = 2
-// eslint-disable-next-line @stylistic/ts/space-infix-ops
-const _c = a-(b -a -b)
+ 
+const _c = a - (b - a - b)
```
Comment (antfu) [inline test/assets/variant-group.vue:5]: Please revert this

File: packages/vscode/package.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
   "publisher": "antfu",
-  "name": "@unocss/vscode",
+  "name": "unocss",
```
Comment (antfu) [inline packages/vscode/package.json:3]: This should only be changed in dev, should not be committed

File: packages/vscode/src/index.ts [hunk 7]
```diff
@@ -1,6 +1,9 @@
 import path from 'path'
+import { watch } from 'fs'
```
Comment (antfu) [inline packages/vscode/src/index.ts:2]: Better use `createFileSystemWatcher` https://code.visualstudio.com/api/references/vscode-api#workspace

File: packages/vscode/src/index.ts [hunk 8]
```diff
@@ -44,10 +46,11 @@ export async function activate(ext: ExtensionContext) {
   status.text = 'UnoCSS'
 
   const root = config.get<string | string[]>('root')
+  const exclude = config.get<string | string[]>('exclude')
+  const include = config.get<string | string[]>('include')
```
Comment (antfu) [inline packages/vscode/src/index.ts:50]: We need to add these two configs in package.json

**Review by antfu — CHANGES_REQUESTED** [review]

File: packages/vscode/src/index.ts (hunk 7 above)
Comment (Simon-He95) [reply packages/vscode/src/index.ts]: Learned

**Comment by Simon-He95** [comment]
Do you think it is necessary for us to monitor the file name changes or new creations of uno.config?

**Review by antfu — APPROVED** [review]

**Comment by antfu** [comment]
Nice work! Thanks

**Pull request merged**